File: solutions/Stack/155_min-stack.py
class MinStack:

    def __init__(self):
        self.stk = []
        self.m = float("inf")

    def push(self, val: int) -> None:
        if not self.stk:
            self.stk.append(0)
            self.m = val
        else:
            x = val - self.m
            self.stk.append(x)
            if x < 0:
                self.m = val
        

    def pop(self) -> None:
        val = self.stk.pop()
        if val < 0:
            self.m -= val

    def top(self) -> int:
        top = self.stk[-1]
        if top > 0: 
            return self.stk[-1] + self.m
        else:
            return self.m
        

    def getMin(self) -> int:
        return self.m
        


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()

#---------------------------------------------------------
#           Time Complexity:
#           push:    O(1)
#           pop:     O(1)
#           top:     O(1)
#           getMin:  O(n)
#           
#           Space Complexity:
#           O(1)
#---------------------------------------------------------


class MinStack:

    def __init__(self):
        self.stk = []
        self.m = []

    def push(self, val: int) -> None:
        self.stk.append(val)
        if not self.m or self.m[-1] >= val:
            self.m.append(val)
        

    def pop(self) -> None:
        val = self.stk.pop()
        if self.m and self.m[-1] == val:
            self.m.pop()
        

    def top(self) -> int:
        return self.stk[-1]
        

    def getMin(self) -> int:
        return self.m[-1]
        

# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()


#---------------------------------------------------------
#           Time Complexity:
#           push:    O(1)
#           pop:     O(1)
#           top:     O(1)
#           getMin:  O(1)
#           
#           Space Complexity:
#           O(n)
#---------------------------------------------------------


# Brute Force Solution
class MinStack:

    def __init__(self):
        self.stk = []

    def push(self, val: int) -> None:
        self.stk.append(val)
        

    def pop(self) -> None:
        self.stk.pop()
        

    def top(self) -> int:
        return self.stk[-1]
        

    def getMin(self) -> int:
        return min(self.stk)
    
#---------------------------------------------------------
#           Time Complexity:
#           push:    O(1)
#           pop:     O(1)
#           top:     O(1)
#           getMin:  O(n)
#           
#           Space Complexity:
#           O(n)
#---------------------------------------------------------

File: C++/Stack/155_min-stack.cpp
#include <stack>

using namespace std;

class MinStack {
private:
    stack<int> stk;
    stack<int> minValue;

public:
    MinStack() {}
    
    void push(int value) {
        stk.push(value);
        if (minValue.empty() || value <= minValue.top()) minValue.push(value);
    }
    
    void pop() {
        if (minValue.top() == stk.top()) minValue.pop();
        stk.pop();
    }
    
    int top() {
        return stk.top();
    }
    
    int getMin() {
        return minValue.top();
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(value);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */



 class MinStack {
private:
    stack<long long> stk;
    long long minValue;

public:
    MinStack() {}
    
    void push(int value) {
        if (stk.empty()) {
            stk.push(0);
            minValue = value;
        } else {
            long long temp = value - minValue;
            stk.push(temp);
            if(temp < 0) minValue = value;
        }
    }
    
    void pop() {
        long long x = stk.top(); stk.pop();
        if (x < 0) minValue -= x;
    }
    
    int top() {
        if (stk.top() < 0) return minValue;
        return stk.top() + minValue;
    }
    
    int getMin() {
        return minValue;
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(value);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */